File: src/codes/scattered-concerns/py/rfc_singleton.py
import os

class Config:
    _instance = None

    def __init__(self):
        self.debug_mode = os.environ.get("DEBUG_MODE") == "true"

    @classmethod
    def get_instance(cls):
        if cls._instance is None:
            cls._instance = Config()
        return cls._instance

    def is_debug(self):
        return self.debug_mode

class Logger:
    _instance = None

    def __init__(self):
        pass

    @classmethod
    def get_instance(cls):
        if cls._instance is None:
            cls._instance = Logger()
        return cls._instance

    def log(self, message: str):
        print(f"[LOG] {message}")

    def debug(self, message: str):
        if Config.get_instance().is_debug():
            print(f"[DEBUG] {message}")

class OrderService:
    def create_order(self, order_id: str):
        logger = Logger.get_instance()
        logger.log(f"注文作成: {order_id}")
        logger.debug(f"注文詳細: {order_id}")

File: src/codes/scattered-concerns/ts/rfc_singleton.ts
class Config {
  private static instance: Config;
  private debugMode: boolean;

  private constructor() {
    this.debugMode = process.env.DEBUG_MODE === "true";
  }

  static getInstance(): Config {
    if (!Config.instance) {
      Config.instance = new Config();
    }
    return Config.instance;
  }

  isDebug(): boolean {
    return this.debugMode;
  }
}

class Logger {
  private static instance: Logger;

  private constructor() {}

  static getInstance(): Logger {
    if (!Logger.instance) {
      Logger.instance = new Logger();
    }
    return Logger.instance;
  }

  log(message: string) {
    console.log(`[LOG] ${message}`);
  }

  debug(message: string) {
    if (Config.getInstance().isDebug()) {
      console.log(`[DEBUG] ${message}`);
    }
  }
}

// 利用例
class OrderService {
  createOrder(orderId: string) {
    const logger = Logger.getInstance();
    logger.log(`注文作成: ${orderId}`);
    logger.debug(`注文詳細: ${orderId}`);
  }
}
